File: bipartite_matching.py
# BOJ 2188 축사 배정

import sys
input = sys.stdin.readline


def dfs(cur):
    vst[cur] = 1
    for nxt in G[cur]:
        if not B[nxt] or (not vst[B[nxt]] and dfs(B[nxt])):
            B[nxt] = cur
            return 1
    return 0


N, M = map(int, input().split())
G = [[] for _ in range(N + 1)]

for i in range(1, N + 1):
    arr = [*map(int, input().split())]
    for j in range(1, arr[0] + 1):
        G[i].append(arr[j])

B = [0] * (M + 1)
ans = 0

for i in range(1, N + 1):
    vst = [0] * (N + 1)
    ans += dfs(i)

print(ans)

File: bipartite_matching.cpp
// BOJ 11375 열혈강호

#include <bits/stdc++.h>
#define sz size()
#define bk back()
#define fi first
#define se second

using namespace std;
typedef long long ll;
typedef pair<int, int> pii;

struct BipartiteMatching {
    int n, m;
    vector<int> b;
    vector<vector<int>> graph;

    BipartiteMatching(int n, int m) : n(n), m(m) {
        graph.resize(n + 1);
        b.resize(m + 1);
    }

    void add_edge(int x, int y) { graph[x].push_back(y); }

    int maximum_matching() {
        int ret = 0;
        vector<bool> vst(n + 1);
        for (int i = 1; i <= n; i++) {
            fill(vst.begin(), vst.end(), false);
            if (dfs(i, vst))
                ret++;
        }

        return ret;
    }

    bool dfs(int cur, vector<bool> &vst) {
        vst[cur] = true;

        for (int nxt : graph[cur]) {
            if (b[nxt] == 0 || (!vst[b[nxt]] && dfs(b[nxt], vst))) {
                b[nxt] = cur;
                return true;
            }
        }

        return false;
    }
};

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int n, m;
    cin >> n >> m;

    BipartiteMatching bm(n, m);
    for (int i = 1; i <= n; i++) {
        int k;
        cin >> k;

        while (k--) {
            int x;
            cin >> x;

            bm.add_edge(i, x);
        }
    }

    cout << bm.maximum_matching();
}
